k2/csrc/self_alignment.cu

```diff
@@ -0,0 +1,241 @@
+/**
+ * Copyright      2022  Xiaomi Corporation (authors: Liyong Guo)
+ *
+ * See LICENSE for clarification regarding multiple authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef K2_CSRC_SELF_ALIGNMENT_H_
+#define K2_CSRC_SELF_ALIGNMENT_H_
+
+#include <torch/extension.h>
+
+#include <vector>
+
+#include "k2/python/csrc/torch.h"
+
+namespace k2 {
+
+FsaVec SelfAlignment(
+    // Normally, ranges is with shape [B][S][T+1] if !modified, [B][S][T] if modified.
+    // Currently, only [B][S][T] is supported.
+    torch::Tensor ranges,
+    torch::Tensor x_lens,  // [B][T]
+    // torch::Tensor blank_connections,
+    torch::Tensor y,
+    // const Ragged<int32_t> &y,
+    torch::Tensor logits,
+    Array1<int32_t> *arc_map) {
+    ContextPtr context;
+    if (ranges.device().type() == torch::kCPU) {
+      context = GetCpuContext();
+    } else if (ranges.is_cuda()) {
+      context = GetCudaContext(ranges.device().index());
+
+      TORCH_CHECK(ranges.get_device() == x_lens.get_device(), "x_lens is on a different device");
+      TORCH_CHECK(ranges.get_device() == y.get_device(), "y device");
+      TORCH_CHECK(ranges.get_device() == logits.get_device(), "logits device");
+
+    } else {
+      K2_LOG(FATAL) << "Unsupported device: " << ranges.device()
+                    << "\nOnly CPU and CUDA are supported";
+    }
+ 
+    TORCH_CHECK(ranges.dim() == 3, "ranges must be 3-dimensional");
+    // U is always 5
+    const int32_t B = ranges.size(0), T = ranges.size(1), U = ranges.size(2);
+
+
+    Dtype t = ScalarTypeToDtype(ranges.scalar_type());
+    K2_CHECK_EQ(torch::kInt, ranges.scalar_type());
+    K2_CHECK_EQ(torch::kInt, x_lens.scalar_type()); // int32_t
+    const float *logits_data = logits.data_ptr<float>();
+    const int32_t *ranges_data = ranges.data_ptr<int32_t>();
```

Comment:
There is another way of accessing these tensors other than extracting the strides etc., which is to use an accessor.
It might be something like:
<code>
 auto range_accessor = ranges.accessor<int32_t, 3>();
</code>
or (and this may be more efficient due to being somehow more specialized or optimize, I'm not sure):
<code>
 auto range_accessor = ranges.packed_accessor32<int32_t, 3>();
</code>
(the 3 is the number of dimensions).  And then I think you'd access it from the lambda as
something like:
<code>
  int32_t i = rng_accessor[idx0][idx1][idx2][idx3];
</code>
as if it were an array.  I think you can also do things like rng_accessor.size(0) to get the 
sizes on each dimension.

HOWEVER: this code may really belong in k2/python/k2/csrc/torch/; this is where we have other code
that makes use of torch APIs, I'm not sure if we really encourage use of torch arrays at this part of the code
(this organization may be a mistake, possibly, but it was intended to separate parts of k2 that
are too deeply entwined with the tensor toolkit).

Reply:
> There is another way of accessing these tensors other than extracting the strides etc., which is to use an accessor.

I think both methods have their advantages and disadvantages.
Now, I don't know which one to choose. would you mind giving me a suggestion on which one should be used?  @danpovey 

For using `accessor`,
according to https://pytorch.org/cppdocs/notes/tensor_basics.html#cuda-accessors,  `It is thus recommended to use accessors for CPU tensors and packed accessors for CUDA tensors.` ( I think the accessor is the reason for the crash when using a CUDA tensor as input.  It's a shame that the unittest I implemented failed to cover this).

To support both cpu and cuda tensor, shall we compile two files pruned_ragged_to_lattice_**cpu**.cu and pruned_ragged_to_lattice_**cuda**.cu (similar to`mutual_information_cuda.cu` and `mutual_information_cpu.cu`)? 

pruned_ragged_to_lattice_**cpu**.cu and pruned_ragged_to_lattice_**cuda**.cu will be very similar except one uses `accessor` and the other uses `packed accessor`.

For using `data_ptr and extracting the strides etc`, 
**only one file(pruned_ragged_to_lattice.cu)** is enough to support both cpu and gpu tensors. But the disadvantage is we could not use [`AT_DISPATCH_FLOATING_TYPES_AND_HALF`  to make it compatible with different input types as 
](https://github.com/k2-fsa/k2/blob/master/k2/python/csrc/torch/pruned_ranges_to_lattice.cu#L180-L182)

```
    // Type of logits can be float32, float64, float16. Though float32 is mainly
    // used, float64 and float16 are also supported for future use.
    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
        logits.scalar_type(), "pruned_ranges_to_lattice", ([&] {
          auto logits_a = logits.accessor<scalar_t, 4>();
```
Since logits.scalar_type() is an enum so it's illegal to define `const scalar_t * logits_data = logits.data_ptr()`.

Reply:
just use packed_accessor32.  I think there is no problem for CPU.